`tilt logs -f` didn't stream logs

### Expected Behavior
Keep streaming logs like tilt web, or `tilt up --stream`

### Current Behavior
`tilt logs -f` just output some logs (like `tilt logs`) then wait there. No more logs stream into output, although the `tilt up --stream` continuously streams logs to output, so does the web. 

### Steps to Reproduce
Tested with an example project, but I couldn't reproduce. Only hit in our project. 
Create an issue here to see if there are any clues or tips on troubleshooting.

### Context
#### `tilt doctor` Output (brew installed)
```
$ tilt doctor
Tilt: v0.19.6, built 2021-04-19
System: darwin-amd64
---
Docker (cluster)
- Host: tcp://127.0.0.1:63700
- Version: 1.41
- Builder: 2
---
Docker (local)
- Host: [default]
- Version: 1.41
- Builder: 2
---
Kubernetes
- Env: minikube
- Context: minikube
- Cluster Name: minikube
- Namespace: default
- Container Runtime: docker
- Version: v1.20.2
- Cluster Local Registry: none
---
```
#### `tilt doctor` Output (bazel downloaded binary, same behavior)
```
$ tilt doctor
Tilt: v0.19.5, built 2021-04-14
System: darwin-amd64
---
Docker (cluster)
- Host: tcp://127.0.0.1:63700
- Version: 1.41
- Builder: 2
---
Docker (local)
- Host: [default]
- Version: 1.41
- Builder: 2
---
Kubernetes
- Env: minikube
- Context: minikube
- Cluster Name: minikube
- Namespace: default
- Container Runtime: docker
- Version: v1.20.2
- Cluster Local Registry: none
---
```

#### About Your Use Case
1. I have a tiltfile with 10+ resources. A few resources generates a lot of runtime logs
2. The environment is started by `tilt up --stream`
3. I want to redirect logs of a certain resource into a separate file continuously, using `tilt logs -f <resource>`. To my surprise it didn't "follow" the log stream as expected. 
4. Meanwhile, I can see logs continuously streaming to web interface and `tilt up --stream` output. 

#### Goroutine dumps of `tilt logs -f`. Hope it provides some clues
```
^\SIGQUIT: quit
PC=0x7fff73060882 m=0 sigcode=0

goroutine 0 [idle]:
runtime.pthread_cond_wait(0x7af13b8, 0x7af1378, 0x0)
	/usr/local/go/src/runtime/sys_darwin.go:384 +0x39
runtime.semasleep(0xffffffffffffffff, 0xc000065000)
	/usr/local/go/src/runtime/os_darwin.go:63 +0x8d
runtime.notesleep(0x7af1170)
	/usr/local/go/src/runtime/lock_sema.go:181 +0xdb
runtime.mPark()
	/usr/local/go/src/runtime/proc.go:1340 +0x39
runtime.stoplockedm()
	/usr/local/go/src/runtime/proc.go:2451 +0x6e
runtime.schedule()
	/usr/local/go/src/runtime/proc.go:3059 +0x48c
runtime.park_m(0xc000cd0780)
	/usr/local/go/src/runtime/proc.go:3274 +0x9d
runtime.mcall(0x4071456)
	/usr/local/go/src/runtime/asm_amd64.s:327 +0x5b

goroutine 1 [select]:
github.com/tilt-dev/tilt/internal/hud/server.(*WebsocketReader).Listen(0xc00024dd50, 0x678c628, 0xc0003d4a00, 0xc000a65b10, 0x0)
	/root/project/internal/hud/server/logs_reader.go:147 +0xf7
github.com/tilt-dev/tilt/internal/hud/server.StreamLogs(0x678c628, 0xc0003d4a00, 0xc0001d3e01, 0x63b857a, 0x2, 0x0, 0x0, 0x0, 0xc000528d9f, 0xf, ...)
	/root/project/internal/hud/server/logs_reader.go:120 +0x3f3
github.com/tilt-dev/tilt/internal/cli.(*logsCmd).run(0xc000269cc9, 0x678c628, 0xc0003d4a00, 0xc000a65b10, 0x0, 0x1, 0x0, 0x0)
	/root/project/internal/cli/logs.go:56 +0x2f0
github.com/tilt-dev/tilt/internal/cli.addCommand.func1(0xc000c362c0, 0xc000a65b10, 0x0, 0x1)
	/root/project/internal/cli/cli.go:146 +0xde
github.com/spf13/cobra.(*Command).execute(0xc000c362c0, 0xc000a65b00, 0x1, 0x1, 0xc000c362c0, 0xc000a65b00)
	/root/project/vendor/github.com/spf13/cobra/command.go:854 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003498c0, 0x7b1fb88, 0x63b9bbc, 0x4)
	/root/project/vendor/github.com/spf13/cobra/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/root/project/vendor/github.com/spf13/cobra/command.go:895
github.com/tilt-dev/tilt/internal/cli.Execute()
	/root/project/internal/cli/cli.go:83 +0x933
main.main()
	/root/project/cmd/tilt/main.go:19 +0xaa

goroutine 19 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x7aefdc0)
	/root/project/vendor/k8s.io/klog/v2/klog.go:1164 +0x8b
created by k8s.io/klog/v2.init.0
	/root/project/vendor/k8s.io/klog/v2/klog.go:418 +0xdf

goroutine 44 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc00005f080)
	/root/project/vendor/go.opencensus.io/stats/view/worker.go:298 +0xcd
created by go.opencensus.io/stats/view.init.0
	/root/project/vendor/go.opencensus.io/stats/view/worker.go:34 +0x68

goroutine 45 [sleep]:
time.Sleep(0x6fc23ac00)
	/usr/local/go/src/runtime/time.go:193 +0xd2
sigs.k8s.io/controller-runtime/pkg/log.init.0.func1()
	/root/project/vendor/sigs.k8s.io/controller-runtime/pkg/log/log.go:63 +0x45
created by sigs.k8s.io/controller-runtime/pkg/log.init.0
	/root/project/vendor/sigs.k8s.io/controller-runtime/pkg/log/log.go:62 +0x35

goroutine 108 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:165 +0x9d
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x25
created by os/signal.Notify.func1.1
	/usr/local/go/src/os/signal/signal.go:151 +0x45

goroutine 109 [chan receive]:
github.com/tilt-dev/tilt/internal/cli.preCommand.func1(0xc000c495c0, 0xc000a65d80, 0x67b0330, 0xc000134d40)
	/root/project/internal/cli/cli.go:122 +0x38
created by github.com/tilt-dev/tilt/internal/cli.preCommand
	/root/project/internal/cli/cli.go:121 +0x40c

goroutine 147 [IO wait]:
internal/poll.runtime_pollWait(0x12084bd0, 0x72, 0xffffffffffffffff)
	/usr/local/go/src/runtime/netpoll.go:222 +0x55
internal/poll.(*pollDesc).wait(0xc000a9f498, 0x72, 0x1000, 0x1000, 0xffffffffffffffff)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc000a9f480, 0xc00017e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:166 +0x1d5
net.(*netFD).Read(0xc000a9f480, 0xc00017e000, 0x1000, 0x1000, 0x672d900, 0xc00000eac8, 0xc001423cd0)
	/usr/local/go/src/net/fd_posix.go:55 +0x4f
net.(*conn).Read(0xc000011440, 0xc00017e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:183 +0x91
bufio.(*Reader).fill(0xc000a350e0)
	/usr/local/go/src/bufio/bufio.go:101 +0x108
bufio.(*Reader).Peek(0xc000a350e0, 0x2, 0xc001423db0, 0x408057a, 0x6735340, 0x7b1f7e0, 0x6730900)
	/usr/local/go/src/bufio/bufio.go:139 +0x4f
github.com/gorilla/websocket.(*Conn).read(0xc000538840, 0x2, 0x672d420, 0xc000a350e0, 0x1, 0x1, 0x0)
	/root/project/vendor/github.com/gorilla/websocket/conn.go:370 +0x46
github.com/gorilla/websocket.(*Conn).advanceFrame(0xc000538840, 0x0, 0x0, 0x0)
	/root/project/vendor/github.com/gorilla/websocket/conn.go:798 +0x5c
github.com/gorilla/websocket.(*Conn).NextReader(0xc000538840, 0x678c628, 0xc0003d4a00, 0x11d0f148, 0xc001df8008, 0x0)
	/root/project/vendor/github.com/gorilla/websocket/conn.go:980 +0x8f
github.com/tilt-dev/tilt/internal/hud/server.(*WebsocketReader).Listen.func1(0xc00082afc0, 0xc00024dd50, 0x678c628, 0xc0003d4a00)
	/root/project/internal/hud/server/logs_reader.go:128 +0x65
created by github.com/tilt-dev/tilt/internal/hud/server.(*WebsocketReader).Listen
	/root/project/internal/hud/server/logs_reader.go:125 +0x91

goroutine 162 [IO wait]:
internal/poll.runtime_pollWait(0x12084cb8, 0x72, 0xffffffffffffffff)
	/usr/local/go/src/runtime/netpoll.go:222 +0x55
internal/poll.(*pollDesc).wait(0xc00005f518, 0x72, 0xf00, 0xfa8, 0xffffffffffffffff)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc00005f500, 0xc000eaa000, 0xfa8, 0xfa8, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:166 +0x1d5
net.(*netFD).Read(0xc00005f500, 0xc000eaa000, 0xfa8, 0xfa8, 0xc001c90000, 0xc000e9b860, 0x413f115)
	/usr/local/go/src/net/fd_posix.go:55 +0x4f
net.(*conn).Read(0xc000011448, 0xc000eaa000, 0xfa8, 0xfa8, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:183 +0x91
crypto/tls.(*atLeastReader).Read(0xc0005fe018, 0xc000eaa000, 0xfa8, 0xfa8, 0x406fe16, 0xc0005a0480, 0xc000cd0d80)
	/usr/local/go/src/crypto/tls/conn.go:776 +0x63
bytes.(*Buffer).ReadFrom(0xc000ea0278, 0x672d660, 0xc0005fe018, 0x400c885, 0x606bde0, 0x631a780)
	/usr/local/go/src/bytes/buffer.go:204 +0xbe
crypto/tls.(*Conn).readFromUntil(0xc000ea0000, 0x1209f1f8, 0xc000011448, 0x5, 0xc000011448, 0x27)
	/usr/local/go/src/crypto/tls/conn.go:798 +0xf3
crypto/tls.(*Conn).readRecordOrCCS(0xc000ea0000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:605 +0x115
crypto/tls.(*Conn).readRecord(...)
	/usr/local/go/src/crypto/tls/conn.go:573
crypto/tls.(*Conn).Read(0xc000ea0000, 0xc000c16000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:1276 +0x165
bufio.(*Reader).Read(0xc0004c21e0, 0xc0012649d8, 0x9, 0x9, 0x11, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:227 +0x222
io.ReadAtLeast(0x672d420, 0xc0004c21e0, 0xc0012649d8, 0x9, 0x9, 0x9, 0xc000e9bd08, 0x41a433b, 0x80c001c80000)
	/usr/local/go/src/io/io.go:328 +0x87
io.ReadFull(...)
	/usr/local/go/src/io/io.go:347
net/http.http2readFrameHeader(0xc0012649d8, 0x9, 0x9, 0x672d420, 0xc0004c21e0, 0x0, 0x0, 0xc000cd1de8, 0x0)
	/usr/local/go/src/net/http/h2_bundle.go:1477 +0x89
net/http.(*http2Framer).ReadFrame(0xc0012649a0, 0xc0005fc000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/h2_bundle.go:1735 +0xa5
net/http.(*http2clientConnReadLoop).run(0xc000e9bfa8, 0x0, 0x0)
	/usr/local/go/src/net/http/h2_bundle.go:8322 +0xd8
net/http.(*http2ClientConn).readLoop(0xc000cd1c80)
	/usr/local/go/src/net/http/h2_bundle.go:8244 +0x6f
created by net/http.(*http2Transport).newClientConn
	/usr/local/go/src/net/http/h2_bundle.go:7208 +0x6c5

rax    0x104
rbx    0x2
rcx    0x7ffeefbff5d8
rdx    0x9000
rdi    0x7af13b8
rsi    0x900100009100
rbp    0x7ffeefbff670
rsp    0x7ffeefbff5d8
r8     0x0
r9     0xa0
r10    0x0
r11    0x202
r12    0x7af13b8
r13    0x16
r14    0x900100009100
r15    0x17874dc0
rip    0x7fff73060882
rflags 0x203
cs     0x7
fs     0x0
gs     0x0
```

When trying `tilt logs -f` in both interactive shell cases and redirecting stdout to another file I saw same behavior. Eventually will be part of larger integration test script - but there's another issue like below (I am sure the "resource" is correct)
```
$ tilt logs -f "resource"
panic: runtime error: index out of range [-1]

goroutine 138 [running]:
github.com/tilt-dev/tilt/pkg/model/logstore.(*LogStore).prevIndexMatchingManifests(...)
        /root/project/pkg/model/logstore/logstore.go:203
github.com/tilt-dev/tilt/pkg/model/logstore.(*LogStore).ContinuingLinesWithOptions(0xc0001bea00, 0x517, 0xc000ba4ab0, 0xc000c51801, 0xc0010e2706, 0x7, 0xc017e2fffb818540)
        /root/project/pkg/model/logstore/logstore.go:451 +0x9cd
github.com/tilt-dev/tilt/internal/hud/server.(*LogStreamer).Handle(0xc0003763a0, 0x0, 0x0, 0xc000572100, 0x16, 0x20, 0x0, 0xc0010d66c0, 0x0, 0xc0010e40c0, ...)
        /root/project/internal/hud/server/logs_reader.go:97 +0x234
github.com/tilt-dev/tilt/internal/hud/server.(*WebsocketReader).handleTextMessage(0xc0004ae9a0, 0x678c628, 0xc00004f340, 0x103ada78, 0xc000c07600, 0x0, 0x0)
        /root/project/internal/hud/server/logs_reader.go:168 +0x13b
github.com/tilt-dev/tilt/internal/hud/server.(*WebsocketReader).Listen.func1(0xc0000c52c0, 0xc0004ae9a0, 0x678c628, 0xc00004f340)
        /root/project/internal/hud/server/logs_reader.go:134 +0xc5
created by github.com/tilt-dev/tilt/internal/hud/server.(*WebsocketReader).Listen
        /root/project/internal/hud/server/logs_reader.go:125 +0x91```

The logs are pretty normal. Just some are in JSON format - again everything is fine in hud, web, or `tilt up --stream`

--- a/internal/hud/server/logs_reader.go
+++ b/internal/hud/server/logs_reader.go
@@ -52,10 +52,17 @@ type ViewHandler interface {
 }
 
 type LogStreamer struct {
-	logstore   *logstore.LogStore
+	logstore *logstore.LogStore
+	// checkpoint tracks the client's latest printed logs.
+	//
+	// WARNING: The server watermark values CANNOT be used for checkpointing within the client!
 	checkpoint logstore.Checkpoint
-	resources  model.ManifestNameSet // if present, resource(s) to stream logs for
-	printer    *hud.IncrementalPrinter
+	// serverWatermark ensures that we don't print any duplicate logs.
+	//
+	// This value should only be used to compare to other server values, NOT client checkpoints.
+	serverWatermark int32
+	resources       model.ManifestNameSet // if present, resource(s) to stream logs for
+	printer         *hud.IncrementalPrinter
 }
 
 func NewLogStreamer(resources []string, p *hud.IncrementalPrinter) *LogStreamer {
@@ -72,20 +79,18 @@ func NewLogStreamer(resources []string, p *hud.IncrementalPrinter) *LogStreamer 
 }
 
 func (ls *LogStreamer) Handle(v proto_webview.View) error {
-	// if printing logs for only one resource, don't need resource name prefix
-	suppressPrefix := len(ls.resources) == 1
-	fromCheckpoint := logstore.Checkpoint(v.LogList.FromCheckpoint)
-	toCheckpoint := logstore.Checkpoint(v.LogList.ToCheckpoint)
-
-	if fromCheckpoint == -1 {
+	if v.LogList.FromCheckpoint == -1 {
 		// Server has no new logs to send
 		return nil
 	}
 
+	// if printing logs for only one resource, don't need resource name prefix
+	suppressPrefix := len(ls.resources) == 1
+
 	segments := v.LogList.Segments
-	if fromCheckpoint < ls.checkpoint {
+	if v.LogList.FromCheckpoint < ls.serverWatermark {
 		// The server is re-sending some logs we already have, so slice them off.
-		deleteCount := ls.checkpoint - fromCheckpoint
+		deleteCount := ls.serverWatermark - v.LogList.FromCheckpoint
 		segments = segments[deleteCount:]
 	}
 
@@ -99,9 +104,8 @@ func (ls *LogStreamer) Handle(v proto_webview.View) error {
 		SuppressPrefix: suppressPrefix,
 	}))
 
-	if toCheckpoint > ls.checkpoint {
-		ls.checkpoint = toCheckpoint
-	}
+	ls.checkpoint = ls.logstore.Checkpoint()
+	ls.serverWatermark = v.LogList.ToCheckpoint
 
 	return nil
 }

--- a/internal/hud/server/logs_reader_test.go
+++ b/internal/hud/server/logs_reader_test.go
@@ -63,22 +63,28 @@ func TestLogStreamerDoesNotPrintResentLogs(t *testing.T) {
 }
 
 func TestLogStreamerCheckpointHandling(t *testing.T) {
-	f := newLogStreamerFixture(t)
-	resourceNames := []string{
-		"foo", "", "foo", "bar",
-		"bar", "bar", "", "bar",
-		"", "foo", "bar", "baz"}
-	view := f.newViewWithLogsForManifests(alphabet[:4], resourceNames[:4], 0)
-	f.handle(view)
+	// the client might attach after the server has truncated, so the first "FromCheckpoint" it sees
+	// won't always be 0
+	for _, initialOffset := range []int32{0, 1000} {
+		t.Run(fmt.Sprintf("Offset-%d", initialOffset), func(t *testing.T) {
+			f := newLogStreamerFixture(t)
+			resourceNames := []string{
+				"foo", "", "foo", "bar",
+				"bar", "bar", "", "bar",
+				"", "foo", "bar", "baz"}
+			view := f.newViewWithLogsForManifests(alphabet[:4], resourceNames[:4], initialOffset)
+			f.handle(view)
 
-	view = f.newViewWithLogsForManifests(alphabet[4:8], resourceNames[4:8], view.LogList.ToCheckpoint)
-	f.handle(view)
+			view = f.newViewWithLogsForManifests(alphabet[4:8], resourceNames[4:8], view.LogList.ToCheckpoint)
+			f.handle(view)
 
-	view = f.newViewWithLogsForManifests(alphabet[8:12], resourceNames[8:12], view.LogList.ToCheckpoint)
-	f.handle(view)
+			view = f.newViewWithLogsForManifests(alphabet[8:12], resourceNames[8:12], view.LogList.ToCheckpoint)
+			f.handle(view)
 
-	expected := f.expectedLinesWithPrefixes(alphabet[:12], resourceNames[:12])
-	f.assertExpectedLogLines(expected)
+			expected := f.expectedLinesWithPrefixes(alphabet[:12], resourceNames[:12])
+			f.assertExpectedLogLines(expected)
+		})
+	}
 }
 
 func TestLogStreamerFiltersOnResourceNamesSingle(t *testing.T) {

